core/server/common/src/main/java/alluxio/master/journal/raft/BufferedJournalApplier.java

```diff
@@ -0,0 +1,295 @@
+          "Catch-up task in progress.");
+      Preconditions.checkState(sequence >= 0, String.format("Invalid sequence: %d", sequence));
+      Preconditions.checkState(mLastAppliedSequence <= sequence, String.format(
+          "Can't catchup to past. Current: %d, Requested: %d", mLastAppliedSequence, sequence));
+      LOG.info("Catching up state machine to sequence: {}", sequence);
+
+      // Complete the request if already at target sequence.
+      if (mLastAppliedSequence == sequence) {
+        return CatchupFuture.completed();
+      }
+
+      // Create an async task for catching up to target sequence.
+      mCatchupThread = new RaftJournalCatchupThread(sequence);
+      mCatchupThread.start();
+      return new CatchupFuture(mCatchupThread);
+    }
+  }
+
+  /**
+   * Applies the entry to master and updates last applied sequence.
+   * Calls to it should be serialized.
+   */
+  private void applyToMaster(Journal.JournalEntry entry) {
+    String masterName;
+    try {
+      masterName = JournalEntryAssociation.getMasterForEntry(entry);
+    } catch (Throwable t) {
+      ProcessUtils.fatalError(LOG, t, "Unrecognized journal entry: %s", entry);
+      throw new IllegalStateException();
+    }
+    try {
+      Journaled master = mJournals.get(masterName).getStateMachine();
+      LOG.trace("Applying entry to master {}: {} ", masterName, entry);
+      master.processJournalEntry(entry);
+      JournalUtils.sinkAppend(mJournalSinks, entry);
+    } catch (Throwable t) {
+      JournalUtils.handleJournalReplayFailure(LOG, t,
+          "Failed to apply journal entry to master %s. Entry: %s", masterName, entry);
+    }
+    // Store last applied sequence.
+    mLastAppliedSequence = entry.getSequenceNumber();
+  }
+
+  /**
+   * RAFT implementation for {@link AbstractCatchupThread}.
+   */
+  class RaftJournalCatchupThread extends AbstractCatchupThread {
+    /** Where to stop catching up. */
+    private long mCatchUpEndSequence;
+    /** Used to stop catching up early. */
+    private boolean mStopCatchingUp = false;
+
+    /**
+     * Creates RAFT journal catch-up thread.
+     *
+     * @param sequence end sequence(inclusive)
+     */
+    public RaftJournalCatchupThread(long sequence) {
+      mCatchUpEndSequence = sequence;
+      setName(String.format("raft-catchup-thread"));
+    }
+
+    @Override
+    public void cancel() {
+      // Signal thread to bail early.
+      synchronized (mSuspendBuffer) {
+        mStopCatchingUp = true;
+        mSuspendBuffer.notifyAll();
+      }
+    }
+
+    protected void runCatchup() {
+      // Spin for catching up until cancelled.
+      while (!mStopCatchingUp && mLastAppliedSequence < mCatchUpEndSequence) {
+        // Wait until notified for cancellation or more entries.
+        synchronized (mSuspendBuffer) {
+          while (!mStopCatchingUp && mSuspendBuffer.size() == 0) {
+            try {
+              mSuspendBuffer.wait();
```

Comment:
I am not sure how `ConcurrentLinkedQueue` work with `Synchronized`.
When running catchup, we 
```
        synchronized (mSuspendBuffer) {
          while (!mStopCatchingUp && mSuspendBuffer.size() == 0) {
            try {
              mSuspendBuffer.wait();
            } catch (InterruptedException e) {
              throw new RuntimeException("Interrupted while catching up.");
            }
          }
```
In this situation, will other `mSuspendBuffer.notifyAll()` operations be blocked by not able to get the `synchronized (mSuspendBuffer)`?

Reply:
Buffer could be accessed by `processJournalEntry`, `resume`, `catchup`.
`processJournalEntry` can be called concurrently with `resume` OR `catchup`. All three won't run in parallel since `resume` first waits until active `catchup` is done.

For when `processJournalEntry` and `catchup` is running concurrently. A monitor is required to get notified when the queue has an entry, so that `catchup` can continue. Note that resume locks only to check if there are entries, so the monitor won't get that hot. For the monitor I used the buffer itself, which will have no effect on concurrency of queue puts/gets.

For when, `processJournalEntry` and `resume` is running concurrently. We don't want to block anything until `resume` decides that it's time to block `processJournalEntry` and finish resuming.